// path-framework/dist/esm2015/path-framework/page/element/button/back-button.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output } from "@angular/core";
import { StringUtility } from "../../../utility/string-utility";
import { Button } from "./button.component";
let BackButtonComponent = class BackButtonComponent {
};
tslib_1.__decorate([
    Input("button"),
    Output("button")
], BackButtonComponent.prototype, "button", void 0);
BackButtonComponent = tslib_1.__decorate([
    Component({
        selector: "path-back-button",
        template: "<a (click)=\"button.onClick()\" class=\"fa-links\">\r\n\t<div class=\"thumbnail tile tile-button {{(button.width <= 1 ? 'tile-medium' : (button.width == 2 ? 'tile-wide' : (button.width == 3 ? 'tile-wide-3' : (button.width >= 4 ? 'tile-wide-4' : ''))))}} backbutton {{button.cssGroupBorder}} {{button.cssClass}}\" [ngStyle]=\"button.cssStyle\">\r\n\t\t<div class=\"h3\">{{button.name}}</div>\r\n\t\t<i class=\"fa fa-2x fa-backward\"></i>\r\n\t</div>\r\n</a>"
    })
], BackButtonComponent);
export { BackButtonComponent };
export class BackButton extends Button {
    constructor(app, pathService, translationService) {
        super(app, pathService, translationService);
    }
    onClick() {
        this.app.navigateBack();
    }
    fromJson(modelElement) {
        if (StringUtility.isEmpty(modelElement["color"])) {
            modelElement["color"] = "silver";
        }
        super.fromJson(modelElement);
        this.type = "backbutton";
        this.name = this.translationService.getText("Back");
    }
}
