// path-framework/dist/esm2015/path-framework/page/element/button/link-button.component.js
import * as tslib_1 from "tslib";
import { Button } from "./button.component";
import { Component, Input, Output } from "@angular/core";
import { KeyUtility } from "../../../utility/key-utility";
let LinkButtonComponent = class LinkButtonComponent {
};
tslib_1.__decorate([
    Input("button"),
    Output("button")
], LinkButtonComponent.prototype, "button", void 0);
LinkButtonComponent = tslib_1.__decorate([
    Component({
        selector: "path-link-button",
        template: "<a (click)=\"button.onClick()\" class=\"fa-links\">\r\n\t<div class=\"thumbnail tile tile-button {{(button.width <= 1 ? 'tile-medium' : (button.width == 2 ? 'tile-wide' : (button.width == 3 ? 'tile-wide-3' : (button.width >= 4 ? 'tile-wide-4' : ''))))}} col-md-3 {{button.cssGroupBorder}} {{button.cssClass}}\" [ngStyle]=\"button.cssStyle\">\r\n\t\t<div class=\"h3\">{{button.name}}</div>\r\n\t\t<i class=\"fa fa-2x {{button.icon}}\"></i>\r\n\t</div>\r\n</a>\r\n\r\n\r\n"
    })
], LinkButtonComponent);
export { LinkButtonComponent };
export class LinkButton extends Button {
    constructor(app, pathService, translationService) {
        super(app, pathService, translationService);
    }
    onClick() {
        let url = null;
        if (this.url.startsWith("/")) {
            // local relative url
            url = this.getApp().getBackendUrl() + this.url + "?token=" + sessionStorage.getItem("pathAppId");
        }
        else {
            // remote absolute url
            url = this.url;
        }
        window.location.assign(url);
    }
    fromJson(modelElement) {
        super.fromJson(modelElement);
        this.type = "linkButton";
        this.url = KeyUtility.translateUrl(modelElement["url"], null, false, this.parentPageElement);
    }
}
