// path-framework/dist/esm2015/path-framework/page/element/button/page-delete-button.component.js
import * as tslib_1 from "tslib";
import { Button } from "./button.component";
import { Output, Input, Component } from "@angular/core";
import { KeyUtility } from "../../../utility/key-utility";
let PageDeleteButtonComponent = class PageDeleteButtonComponent {
};
tslib_1.__decorate([
    Input("button"),
    Output("button")
], PageDeleteButtonComponent.prototype, "button", void 0);
PageDeleteButtonComponent = tslib_1.__decorate([
    Component({
        selector: "path-page-delete-button",
        template: "<a (click)=\"button.onClick()\" class=\"fa-links\">\r\n\t<div class=\"thumbnail tile tile-button {{(button.width <= 1 ? 'tile-medium' : (button.width == 2 ? 'tile-wide' : (button.width == 3 ? 'tile-wide-3' : (button.width >= 4 ? 'tile-wide-4' : ''))))}} tile-red col-md-3 {{button.cssGroupBorder}} {{button.cssClass}}\" [ngStyle]=\"button.cssStyle\">\r\n\t\t<div class=\"h3\">{{button.name}}</div>\r\n\t\t<i class=\"fa fa-2x fa-remove\"></i>\r\n\t</div>\r\n</a>\r\n"
    })
], PageDeleteButtonComponent);
export { PageDeleteButtonComponent };
export class PageDeleteButton extends Button {
    constructor(app, pathService, translationService) {
        super(app, pathService, translationService);
        this.name = translationService.getText("Delete");
    }
    onClick() {
        this.getApp().yesNo(this.translationService.getText("DeleteWarningQuestion"), () => {
            this.pathService.serverDelete(this.getApp().getBackendUrl(), this.url, () => {
                this.getApp().navigateBack();
                this.getApp().refreshCurrentPage();
            });
        }, () => {
        });
    }
    fromJson(modelElement) {
        super.fromJson(modelElement);
        this.type = "deleteButton";
        this.url = KeyUtility.translateUrl(modelElement["url"], null, false, this.parentPageElement);
    }
}
