// path-framework/dist/esm2015/path-framework/page/element/custom/custom-container.component.js
import * as tslib_1 from "tslib";
import { Component, ComponentFactoryResolver, Input, OnDestroy, OnInit, ViewChild } from "@angular/core";
import { CustomDirective } from "./custom.directive";
let CustomContainerComponent = class CustomContainerComponent {
    constructor(componentFactoryResolver) {
        this.componentFactoryResolver = componentFactoryResolver;
    }
    ngOnInit() {
        this.loadComponent();
    }
    ngOnDestroy() {
    }
    loadComponent() {
        const componentFactory = this.componentFactoryResolver.resolveComponentFactory(this.pageElement.typeClass);
        const viewContainerRef = this.customDirective.viewContainerRef;
        viewContainerRef.clear();
        const componentRef = viewContainerRef.createComponent(componentFactory);
        componentRef.instance.pageElement = this.pageElement;
    }
};
CustomContainerComponent.ctorParameters = () => [
    { type: ComponentFactoryResolver }
];
tslib_1.__decorate([
    Input()
], CustomContainerComponent.prototype, "pageElement", void 0);
tslib_1.__decorate([
    ViewChild(CustomDirective, { static: true })
], CustomContainerComponent.prototype, "customDirective", void 0);
CustomContainerComponent = tslib_1.__decorate([
    Component({
        selector: "path-custom-container-component",
        template: `
        <ng-template path-custom-directive></ng-template>`
    })
], CustomContainerComponent);
export { CustomContainerComponent };
export class CustomPageElement {
    get pageElement() {
        return this._pageElement;
    }
    set pageElement(value) {
        this._pageElement = value;
    }
}
