// path-framework/tests/path-framework/TestAppComponent.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { Component } from "@angular/core";
import { PathAppComponent } from "../../app/path-framework/path-app.component";
import { PathService } from "../../app/path-framework/service/path.service";
import { TranslationService } from "../../app/path-framework/service/translation.service";
var TestAppComponent = /** @class */ (function (_super) {
    __extends(TestAppComponent, _super);
    function TestAppComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TestAppComponent.prototype.getBackendUrl = function () {
        return "";
    };
    TestAppComponent.prototype.getBeans = function () {
    };
    TestAppComponent.prototype.getFrontendVersion = function () {
        return "";
    };
    TestAppComponent.prototype.getGuiModel = function () {
    };
    TestAppComponent.prototype.getHandlers = function () {
    };
    TestAppComponent.prototype.getOwnUserForm = function () {
        return "";
    };
    TestAppComponent.prototype.getStartPage = function () {
        return "";
    };
    TestAppComponent = __decorate([
        Component({
            selector: "path-application",
            template: "<span></span>",
            providers: [PathService, TranslationService]
        })
    ], TestAppComponent);
    return TestAppComponent;
}(PathAppComponent));
export { TestAppComponent };
