// path-framework/dist/esm2015/path-framework/form/field/form-field.js
export class FormField {
    constructor(form, translationService) {
        this.form = form;
        this.translationService = translationService;
    }
    getForm() {
        return this.form;
    }
    get height() {
        return this._height;
    }
    set height(value) {
        this._height = value;
    }
    get visible() {
        return this._visible;
    }
    set visible(value) {
        this._visible = value;
    }
    get width() {
        return this._width;
    }
    set width(value) {
        this._width = value;
    }
    get newRow() {
        return this._newRow;
    }
    set newRow(value) {
        this._newRow = value;
    }
    get required() {
        return this._required;
    }
    set required(value) {
        this._required = value;
    }
    get readonly() {
        return this._readonly;
    }
    set readonly(value) {
        this._readonly = value;
    }
    get id() {
        return this._id;
    }
    set id(value) {
        this._id = value;
    }
    get name() {
        return this._name;
    }
    set name(value) {
        this._name = value;
    }
    get type() {
        return this._type;
    }
    set type(value) {
        this._type = value;
    }
    get labelVisible() {
        return this._labelVisible;
    }
    set labelVisible(value) {
        this._labelVisible = value;
    }
    isReadonly() {
        return this.readonly;
    }
    fromJson(modelFormField) {
        this.visible = true;
        if (modelFormField["visible"] != null) {
            this.visible = modelFormField["visible"];
        }
        this.labelVisible = true;
        if (modelFormField["labelVisible"] != null) {
            this.labelVisible = modelFormField["labelVisible"];
        }
        this.width = 1;
        if (modelFormField["width"] != null) {
            this.width = modelFormField["width"];
        }
        this.newRow = false;
        if (modelFormField["newRow"] != null) {
            this.newRow = modelFormField["newRow"];
        }
        this.required = false;
        if (modelFormField["required"] != null) {
            this.required = modelFormField["required"];
        }
        this.readonly = false;
        if (modelFormField["readonly"] != null) {
            this.readonly = modelFormField["readonly"];
        }
        this.id = modelFormField["id"];
        this.name = this.translationService.getText(modelFormField.name);
        this.type = modelFormField.type;
        this.height = modelFormField["height"];
    }
}
