// path-framework/dist/esm5/path-framework/form/field/date/date-field.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output } from "@angular/core";
import { ValueField } from "../value-field";
import * as momentImported from "moment";
var moment = momentImported;
var DateFieldComponent = /** @class */ (function () {
    function DateFieldComponent() {
    }
    tslib_1.__decorate([
        Input("field"),
        Output("field")
    ], DateFieldComponent.prototype, "field", void 0);
    DateFieldComponent = tslib_1.__decorate([
        Component({
            selector: "path-datefield",
            template: "<path-form-field-label [label]=\"field.name\" [required]=\"field.required\"></path-form-field-label>\r\n<div [class.col-md-4]=\"field.width == 1\" [class.col-md-10]=\"field.width == 2\">\r\n    <div class=\"input-group\">\r\n\t    <span class=\"input-group-btn\">\r\n\t\t    <button (click)=\"datepicker.toggle()\" class=\"btn btn-default\" type=\"button\"><i class=\"fa fa-calendar fa-fw\"></i></button>\r\n\t    </span>\r\n\t    <input #datepicker=\"bsDatepicker\" #model=\"ngModel\" bsDatepicker [bsConfig]=\"{ dateInputFormat: field.dateInputFormat, containerClass: 'theme-default' }\" placement=\"bottom left\" [bsValue]=\"field.datePickerInitialValue\" [(ngModel)]=\"field.datePickerValue\" (bsValueChange)=\"field.updateValueFromGui($event)\" [isDisabled]=\"field.isReadonly()\" [readonly]=\"field.isReadonly()\" [class.required-field]=\"(!model.valid &&!model.pristine && field.required)\" [required]=\"field.required\" style=\"display: block; width: 100%\" type=\"text\" class=\"form-control\" placeholder=\"\">\r\n    </div>\r\n</div>\r\n"
        })
    ], DateFieldComponent);
    return DateFieldComponent;
}());
export { DateFieldComponent };
var DateField = /** @class */ (function (_super) {
    tslib_1.__extends(DateField, _super);
    function DateField(form, translationService) {
        var _this = _super.call(this, form, translationService) || this;
        _this.form = form;
        _this.translationService = translationService;
        _this._initialValueSet = false;
        _this._dateInputFormat = translationService.getUserDateFormat();
        return _this;
    }
    Object.defineProperty(DateField.prototype, "datePickerValue", {
        get: function () {
            return this._datePickerValue;
        },
        set: function (value) {
            this._datePickerValue = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DateField.prototype, "initialValueSet", {
        get: function () {
            return this._initialValueSet;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DateField.prototype, "datePickerInitialValue", {
        get: function () {
            return this._datePickerInitialValue;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DateField.prototype, "dateInputFormat", {
        get: function () {
            return this._dateInputFormat;
        },
        enumerable: true,
        configurable: true
    });
    DateField.prototype.setValue = function (value) {
        if (typeof value === "string") {
            value = moment(value).toDate();
        }
        if (value != null && Object.prototype.toString.call(value) === "[object Date]") {
            // it is a date
            if (isNaN(value.getTime())) {
                value = null;
            }
            else {
                // cut off local timezone
                value = new Date(value.toDateString() + " 00:00:00 GMT");
            }
        }
        else {
            value = null;
        }
        // update ui
        this._datePickerValue = value;
        _super.prototype.setValue.call(this, value);
        if (!this._initialValueSet) {
            this._datePickerInitialValue = value;
            this._initialValueSet = true;
        }
    };
    DateField.prototype.updateValueFromGui = function (value) {
        if (value != null && isNaN(value.getDate())) {
            value = null;
        }
        if (value != null) {
            this.setValue(value);
            this._datePickerValue = this.value;
        }
    };
    return DateField;
}(ValueField));
export { DateField };
