// path-framework/dist/esm2015/path-framework/page/page.js
import { PageElement } from "./element/page-element";
export class Page {
    constructor(parentPageElement) {
        this._content = [];
        this._rows = [];
        this._parentPageElement = parentPageElement;
    }
    getShortName() {
        return PageElement.buildShortName(this.name);
    }
    get name() {
        return this._name;
    }
    set name(value) {
        this._name = value;
    }
    get content() {
        return this._content;
    }
    set content(value) {
        this._content = value;
    }
    get id() {
        return this._id;
    }
    set id(value) {
        this._id = value;
    }
    get parentPageElement() {
        return this._parentPageElement;
    }
    updateRows() {
        const rows = [];
        let currentPageRow = null;
        for (const pageElement of this.content) {
            if (pageElement.newRow === true || currentPageRow == null) {
                currentPageRow = new PageRow();
                rows.push(currentPageRow);
            }
            currentPageRow.pageElements.push(pageElement);
        }
        this._rows = rows;
    }
}
class PageRow {
    constructor() {
        this._pageElements = [];
    }
    get pageElements() {
        return this._pageElements;
    }
    isVisible() {
        for (const pageElement of this.pageElements) {
            if (pageElement.visible) {
                return true;
            }
        }
        return false;
    }
}
