// path-framework/dist/esm2015/path-framework/form/form.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output } from "@angular/core";
import { ValueField } from "./field/value-field";
import { FieldListField } from "./field/fieldList/field-list-field.component";
import { FocusUtility } from "./focus-utility";
let FormComponent = class FormComponent {
    onSubmit() {
        this.form.close(true, false);
    }
    ngAfterViewInit() {
        FocusUtility.focusFirstField(this.form);
    }
};
tslib_1.__decorate([
    Input("form"),
    Output("form")
], FormComponent.prototype, "form", void 0);
FormComponent = tslib_1.__decorate([
    Component({
        selector: "path-form",
        template: "<div [ngClass]=\"form.borderStyle == 2 ? 'modal-content' : ''\">\r\n\t<form (ngSubmit)=\"onSubmit()\" #pathForm=\"ngForm\" autocomplete=\"off\" ngNativeValidate>\r\n\t\t<div *ngIf=\"form.headerVisible\" class=\"modal-header\">\r\n\t\t\t<button type=\"button\" class=\"close\" data-dismiss=\"modal\" (click)=\"form.close(false,false)\">&times;</button>\r\n\t\t\t<h5 class=\"modal-title\">{{form.title}}</h5>\r\n\t\t</div>\r\n\t\t<div class=\"modal-body\">\r\n\t\t\t<div *ngFor='let row of form.rows;' class=\"row\" [hidden]=\"!row.isVisible()\">\r\n\t\t\t\t<div *ngFor='let field of row.fields;' [hidden]=\"!field.visible\">\r\n\t\t\t\t\t<div *ngIf=\"field.type == 'text'\">\r\n\t\t\t\t\t\t<path-textfield [field]=\"field\"></path-textfield>\r\n\t\t\t\t\t</div>\r\n\t\t\t\t\t<div *ngIf=\"field.type == 'translation'\">\r\n\t\t\t\t\t\t<path-translationfield [field]=\"field\"></path-translationfield>\r\n\t\t\t\t\t</div>\r\n\t\t\t\t\t<div *ngIf=\"field.type == 'number'\">\r\n\t\t\t\t\t\t<path-numberfield [field]=\"field\"></path-numberfield>\r\n\t\t\t\t\t</div>\r\n\t\t\t\t\t<div *ngIf=\"field.type == 'label'\">\r\n\t\t\t\t\t\t<path-label-field [field]=\"field\"></path-label-field>\r\n\t\t\t\t\t</div>\r\n\t\t\t\t\t<div *ngIf=\"field.type == 'fieldList'\">\r\n\t\t\t\t\t\t<path-field-list [field]=\"field\"></path-field-list>\r\n\t\t\t\t\t</div>\r\n\t\t\t\t\t<div *ngIf=\"field.type == 'date'\">\r\n\t\t\t\t\t\t<path-datefield [field]=\"field\"></path-datefield>\r\n\t\t\t\t\t</div>\r\n\t\t\t\t\t<div *ngIf=\"field.type == 'autocomplete'\">\r\n\t\t\t\t\t\t<path-autocomplete [field]=\"field\"></path-autocomplete>\r\n\t\t\t\t\t</div>\r\n\t\t\t\t\t<div *ngIf=\"field.type == 'ProgressBarField'\">\r\n\t\t\t\t\t\t<path-progress-bar [field]=\"field\"></path-progress-bar>\r\n\t\t\t\t\t</div>\r\n\t\t\t\t\t<div *ngIf=\"field.type == 'RadioGroupField'\">\r\n\t\t\t\t\t\t<path-radio-group [field]=\"field\"></path-radio-group>\r\n\t\t\t\t\t</div>\r\n\t\t\t\t\t<div *ngIf=\"field.type == 'CheckboxGroupField'\">\r\n\t\t\t\t\t\t<path-checkbox-group [field]=\"field\"></path-checkbox-group>\r\n\t\t\t\t\t</div>\r\n\t\t\t\t\t<div *ngIf=\"field.type == 'fileUpload'\">\r\n\t\t\t\t\t\t<path-file-upload [field]=\"field\"></path-file-upload>\r\n\t\t\t\t\t</div>\r\n\t\t\t\t</div>\r\n\t\t\t</div>\r\n\t\t\t<!-- show form buttons on modal-body in mobile version, otherwise they are taking up too much space when the keyboard is shown -->\r\n\t\t\t<div [hidden]=\"!form.footerVisible\" class=\"hide-on-desktop\">\r\n\t\t\t\t<div class=\"row\">\r\n                    <div class=\"col-md-12\">\r\n\t\t\t\t\t<div class=\"col-md-2 buttons-smartphone-left\">\r\n\t\t\t\t\t<span *ngFor='let field of form.fields' class=\"pull-left\" [hidden]=\"!field.visible\">\r\n\t\t\t\t\t\t<span *ngIf=\"field.type == 'deleteButton'\">\r\n\t\t\t\t\t\t\t<input [disabled]=\"field.isReadonly()\" type=\"button\" class=\"btn btn-default btn-footer\" (click)=\"field.onClick()\" value=\"{{field.name}}\"/>\r\n\t\t\t\t\t\t</span>\r\n\t\t\t\t\t\t<span *ngIf=\"field.type == 'previousButton'\">\r\n\t\t\t\t\t\t\t<input [disabled]=\"field.isReadonly()\" type=\"button\" class=\"btn btn-default btn-footer\" (click)=\"field.onClick()\" value=\"{{field.name}}\"/>\r\n\t\t\t\t\t\t</span>\r\n\t\t\t\t\t</span>\r\n\t\t\t\t\t</div>\r\n\t\t\t\t\t<div class=\"col-md-10 buttons-smartphone-right\">\r\n\t\t\t\t\t<span *ngFor='let field of form.fields' [hidden]=\"!field.visible\">\r\n\t\t\t\t\t\t<span *ngIf=\"field.type == 'okButton'\">\r\n\t\t\t\t\t\t\t<input [disabled]=\"field.isReadonly()\" type=\"submit\" class=\"btn btn-default btn-footer\" (click)=\"field.onClick()\" value=\"{{field.name}}\"/>\r\n\t\t\t\t\t\t</span>\r\n\t\t\t\t\t\t<span *ngIf=\"field.type == 'cancelButton'\">\r\n\t\t\t\t\t\t\t<input [disabled]=\"field.isReadonly()\" type=\"button\" class=\"btn btn-default btn-footer\" (click)=\"field.onClick()\" value=\"{{field.name}}\"/>\r\n\t\t\t\t\t\t</span>\r\n\t\t\t\t\t</span>\r\n\t\t\t\t\t</div>\r\n\t\t\t\t</div>\r\n                </div>\r\n\t\t\t</div>\r\n\t\t</div>\r\n\t\t<!-- default version for desktop devices -->\r\n\t\t<div class=\"hide-on-mobile\">\r\n\t\t\t<div [hidden]=\"!form.footerVisible\" class=\"modal-footer\">\r\n\t\t\t<div class=\"row\">\r\n\t\t\t\t<div class=\"col-md-2\">\r\n\t\t\t\t\t\t<span *ngFor='let field of form.fields' class=\"pull-left\" [hidden]=\"!field.visible\">\r\n\t\t\t\t\t\t\t<span *ngIf=\"field.type == 'deleteButton'\">\r\n\t\t\t\t\t\t\t\t<input [disabled]=\"field.isReadonly()\" type=\"button\" class=\"btn btn-default btn-footer\" (click)=\"field.onClick()\" value=\"{{field.name}}\"/>\r\n\t\t\t\t\t\t\t</span>\r\n\t\t\t\t\t\t\t<span *ngIf=\"field.type == 'previousButton'\">\r\n\t\t\t\t\t\t\t\t<input [disabled]=\"field.isReadonly()\" type=\"button\" class=\"btn btn-default btn-footer\" (click)=\"field.onClick()\" value=\"{{field.name}}\"/>\r\n\t\t\t\t\t\t\t</span>\r\n\t\t\t\t\t\t</span>\r\n\t\t\t\t</div>\r\n\t\t\t\t<div class=\"col-md-10\">\r\n\t\t\t\t\t\t<span *ngFor='let field of form.fields' [hidden]=\"!field.visible\">\r\n\t\t\t\t\t\t\t<span *ngIf=\"field.type == 'okButton'\">\r\n\t\t\t\t\t\t\t\t<input [disabled]=\"field.isReadonly()\" type=\"submit\" class=\"btn btn-default btn-footer\" (click)=\"field.onClick()\" value=\"{{field.name}}\"/>\r\n\t\t\t\t\t\t\t</span>\r\n\t\t\t\t\t\t\t<span *ngIf=\"field.type == 'cancelButton'\">\r\n\t\t\t\t\t\t\t\t<input [disabled]=\"field.isReadonly()\" type=\"button\" class=\"btn btn-default btn-footer\" (click)=\"field.onClick()\" value=\"{{field.name}}\"/>\r\n\t\t\t\t\t\t\t</span>\r\n\t\t\t\t\t\t</span>\r\n\t\t\t\t</div>\r\n\t\t\t</div>\r\n\t\t\t</div>\r\n\t\t</div>\r\n\t</form>\r\n</div>"
    })
], FormComponent);
export { FormComponent };
export class Form {
    constructor(pathService, app) {
        this.pathService = pathService;
        this.app = app;
        this._fields = [];
        this._rows = [];
        this._headerVisible = true;
        this._footerVisible = true;
        this._borderStyle = BorderStyle.Shadow;
    }
    getApp() {
        return this.app;
    }
    getKey() {
        return this.key;
    }
    get key() {
        return this._key;
    }
    set key(value) {
        this._key = value;
    }
    get rows() {
        return this._rows;
    }
    get title() {
        return this._title;
    }
    set title(value) {
        this._title = value;
    }
    getFields() {
        return this.fields;
    }
    get fields() {
        return this._fields;
    }
    set fields(value) {
        this._fields = value;
    }
    get handler() {
        return this._handler;
    }
    set handler(value) {
        this._handler = value;
    }
    get url() {
        return this._url;
    }
    set url(value) {
        this._url = value;
    }
    get bean() {
        return this._bean;
    }
    set bean(value) {
        this._bean = value;
    }
    get formFunction() {
        return this._formFunction;
    }
    set formFunction(value) {
        this._formFunction = value;
    }
    get headerVisible() {
        return this._headerVisible;
    }
    set headerVisible(value) {
        this._headerVisible = value;
    }
    get footerVisible() {
        return this._footerVisible;
    }
    set footerVisible(value) {
        this._footerVisible = value;
    }
    get borderStyle() {
        return this._borderStyle;
    }
    set borderStyle(value) {
        this._borderStyle = value;
    }
    get urlDefaults() {
        return this._urlDefaults;
    }
    fromJson(modelForm) {
        this.url = modelForm["url"];
        if (modelForm["urlDefaults"] != null) {
            this._urlDefaults = modelForm["urlDefaults"];
        }
        this.headerVisible = true;
        if (modelForm["headerVisible"] != null) {
            this.headerVisible = modelForm["headerVisible"];
        }
        this.footerVisible = true;
        if (modelForm["footerVisible"] != null) {
            this.footerVisible = modelForm["footerVisible"];
        }
        this.borderStyle = BorderStyle.Shadow;
        if (modelForm["borderStyle"] != null) {
            const borderStyleString = modelForm["borderStyle"];
            this.borderStyle = BorderStyle[borderStyleString];
        }
    }
    updateRows() {
        const rows = [];
        let currentFormRow;
        for (const field of this.fields) {
            if (field instanceof FieldListField) {
                for (const subField of field.subfields) {
                    currentFormRow = this.calculateFieldRow(subField, currentFormRow, rows);
                    currentFormRow.fields.push(subField);
                }
            }
            else {
                currentFormRow = this.calculateFieldRow(field, currentFormRow, rows);
                currentFormRow.fields.push(field);
            }
        }
        this._rows = rows;
    }
    calculateFieldRow(field, currentFormRow, rows) {
        // auto-start new row with form width 2
        if (currentFormRow == null ||
            field.newRow ||
            currentFormRow.fields.length >= 2 ||
            field.width >= 2 ||
            currentFormRow.getWidth() >= 2) {
            field.newRow = true;
            currentFormRow = new FormRow();
            rows.push(currentFormRow);
        }
        return currentFormRow;
    }
    close(save, remove) {
        if (save || remove) {
            // call close handler
            if (this.handler != null) {
                this.handler.doSave(this._bean);
            }
            const data = {};
            for (const field of this._fields) {
                if (field instanceof ValueField && field.id != null) {
                    data[field.id] = field.value;
                }
                if (field instanceof FieldListField) {
                    for (const subfield of field.subfields) {
                        data[subfield.id] = subfield.value;
                    }
                }
            }
            if (remove) {
                this.pathService.serverDelete(this.app.getBackendUrl(), this.url, this.formFunction.delete);
            }
            else if (this.key == null) {
                // create
                this.pathService.serverPost(this.app.getBackendUrl(), this.url, data, this.formFunction.save, null);
            }
            else {
                // update (with key)
                this.pathService.serverPut(this.app.getBackendUrl(), this.url, data, this.formFunction.save);
            }
        }
        else {
            this.formFunction.cancel();
        }
    }
    onKey(event) {
        if (event.keyCode === 27) { // esc
            this.close(false, false);
        }
    }
}
export class FormRow {
    constructor() {
        this._fields = [];
    }
    get fields() {
        return this._fields;
    }
    set fields(value) {
        this._fields = value;
    }
    getWidth() {
        let sum = 0;
        for (const field of this.fields) {
            sum += field.width;
        }
        return sum;
    }
    isVisible() {
        for (const field of this.fields) {
            if (field.visible) {
                return true;
            }
        }
        return false;
    }
}
export var BorderStyle;
(function (BorderStyle) {
    // TODO use string enum after TS upgrade, then use enum in template (with member)
    BorderStyle[BorderStyle["None"] = 1] = "None";
    BorderStyle[BorderStyle["Shadow"] = 2] = "Shadow";
})(BorderStyle || (BorderStyle = {}));
