// path-framework/dist/esm2015/path-framework/form/field/number/number-field.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output } from "@angular/core";
import { ValueField } from "../value-field";
let NumberFieldComponent = class NumberFieldComponent {
};
tslib_1.__decorate([
    Input("field"),
    Output("field")
], NumberFieldComponent.prototype, "field", void 0);
NumberFieldComponent = tslib_1.__decorate([
    Component({
        selector: "path-numberfield",
        template: "<path-form-field-label [label]=\"field.name\" [required]=\"field.required\"></path-form-field-label>\r\n<div [class.col-md-4]=\"field.width == 1\" [class.col-md-10]=\"field.width == 2\">\r\n    <div class=\"input-group\">\r\n        <input min=\"{{field.min}}\" max=\"{{field.max}}\" step=\"{{field.getStep()}}\" *ngIf=\"field.height == null || field.height <= 1\" #model=\"ngModel\" [readonly]=\"field.isReadonly()\" [class.required-field]=\"(!model.valid &&!model.pristine && field.required)\" [required]=\"field.required\" [(ngModel)]=field.value type=\"number\" class=\"form-control\" placeholder=\"\">\r\n    </div>\r\n</div>\r\n"
    })
], NumberFieldComponent);
export { NumberFieldComponent };
export class NumberField extends ValueField {
    get min() {
        return this._min;
    }
    set min(value) {
        this._min = value;
    }
    get max() {
        return this._max;
    }
    set max(value) {
        this._max = value;
    }
    get digits() {
        return this._digits;
    }
    set digits(value) {
        this._digits = value;
    }
    getStep() {
        if (this._digits == null) {
            return "";
        }
        else if (this._digits === 0) {
            return "0";
        }
        // e.g. 3 => 0.001
        let step = "0.";
        for (let k = 0; k < this._digits - 1; k++) {
            step = step + "0";
        }
        return step + "1";
    }
    fromJson(modelFormField) {
        super.fromJson(modelFormField);
        if (modelFormField["min"] != null) {
            this.min = modelFormField["min"];
        }
        if (modelFormField["max"] != null) {
            this.max = modelFormField["max"];
        }
        if (modelFormField["digits"] != null) {
            this.digits = modelFormField["digits"];
        }
    }
}
