// path-framework/dist/esm5/path-framework/form/field/checkbox/checkbox-group.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output } from "@angular/core";
import { ValueField } from "../value-field";
var CheckboxGroupComponent = /** @class */ (function () {
    function CheckboxGroupComponent() {
    }
    tslib_1.__decorate([
        Input("field"),
        Output("field")
    ], CheckboxGroupComponent.prototype, "field", void 0);
    CheckboxGroupComponent = tslib_1.__decorate([
        Component({
            selector: "path-checkbox-group",
            template: "<path-form-field-label [label]=\"field.name\" [required]=\"field.required\"></path-form-field-label>\r\n<div class=\"checkbox\" [class.col-md-4]=\"field.width == 1\" [class.col-md-10]=\"field.width == 2\">\r\n    <span *ngFor=\"let entry of field.entries\">\r\n        <label class=\"checkbox-inline\">\r\n            <input type=\"checkbox\" [(ngModel)]=\"entry.selected\" (ngModelChange)=\"field.doClick()\">\r\n            <span class=\"checkbox-label\"> {{entry.name}}&nbsp;&nbsp;</span>\r\n        </label>\r\n    </span>\r\n</div>\r\n"
        })
    ], CheckboxGroupComponent);
    return CheckboxGroupComponent;
}());
export { CheckboxGroupComponent };
var CheckboxGroupField = /** @class */ (function (_super) {
    tslib_1.__extends(CheckboxGroupField, _super);
    function CheckboxGroupField() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._entries = [];
        return _this;
    }
    Object.defineProperty(CheckboxGroupField.prototype, "entries", {
        get: function () {
            return this._entries;
        },
        set: function (value) {
            this._entries = value;
        },
        enumerable: true,
        configurable: true
    });
    CheckboxGroupField.prototype.doClick = function () {
        var e_1, _a;
        var values = [];
        try {
            for (var _b = tslib_1.__values(this.entries), _c = _b.next(); !_c.done; _c = _b.next()) {
                var entry = _c.value;
                if (entry.selected) {
                    values.push(entry.key);
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        this.setValue(values);
    };
    CheckboxGroupField.prototype.setValue = function (values) {
        var e_2, _a, e_3, _b, e_4, _c;
        _super.prototype.setValue.call(this, values);
        try {
            for (var _d = tslib_1.__values(this.entries), _e = _d.next(); !_e.done; _e = _d.next()) {
                var entry = _e.value;
                entry.selected = false;
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_e && !_e.done && (_a = _d.return)) _a.call(_d);
            }
            finally { if (e_2) throw e_2.error; }
        }
        try {
            for (var values_1 = tslib_1.__values(values), values_1_1 = values_1.next(); !values_1_1.done; values_1_1 = values_1.next()) {
                var value = values_1_1.value;
                try {
                    for (var _f = (e_4 = void 0, tslib_1.__values(this.entries)), _g = _f.next(); !_g.done; _g = _f.next()) {
                        var entry = _g.value;
                        if (entry.key === value) {
                            entry.selected = true;
                        }
                    }
                }
                catch (e_4_1) { e_4 = { error: e_4_1 }; }
                finally {
                    try {
                        if (_g && !_g.done && (_c = _f.return)) _c.call(_f);
                    }
                    finally { if (e_4) throw e_4.error; }
                }
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (values_1_1 && !values_1_1.done && (_b = values_1.return)) _b.call(values_1);
            }
            finally { if (e_3) throw e_3.error; }
        }
    };
    CheckboxGroupField.prototype.fromJson = function (modelFormField) {
        var e_5, _a;
        _super.prototype.fromJson.call(this, modelFormField);
        try {
            for (var _b = tslib_1.__values(modelFormField["data"]), _c = _b.next(); !_c.done; _c = _b.next()) {
                var entryModel = _c.value;
                var entry = new CheckboxGroupEntry();
                entry.name = entryModel.name;
                entry.key = entryModel.key;
                entry.selected = false;
                this.entries.push(entry);
            }
        }
        catch (e_5_1) { e_5 = { error: e_5_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_5) throw e_5.error; }
        }
    };
    return CheckboxGroupField;
}(ValueField));
export { CheckboxGroupField };
var CheckboxGroupEntry = /** @class */ (function () {
    function CheckboxGroupEntry() {
    }
    Object.defineProperty(CheckboxGroupEntry.prototype, "key", {
        get: function () {
            return this._key;
        },
        set: function (value) {
            this._key = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CheckboxGroupEntry.prototype, "name", {
        get: function () {
            return this._name;
        },
        set: function (value) {
            this._name = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CheckboxGroupEntry.prototype, "selected", {
        get: function () {
            return this._selected;
        },
        set: function (value) {
            this._selected = value;
        },
        enumerable: true,
        configurable: true
    });
    return CheckboxGroupEntry;
}());
export { CheckboxGroupEntry };
