// path-framework/dist/esm5/path-framework/form/field/radio/radio-group.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output } from "@angular/core";
import { Radio } from "./radio";
import { ValueField } from "../value-field";
var RadioGroupComponent = /** @class */ (function () {
    function RadioGroupComponent() {
    }
    tslib_1.__decorate([
        Input("field"),
        Output("field")
    ], RadioGroupComponent.prototype, "field", void 0);
    RadioGroupComponent = tslib_1.__decorate([
        Component({
            selector: "path-radio-group",
            template: "<path-form-field-label [label]=\"field.name\" [required]=\"field.required\"></path-form-field-label>\r\n<div [class.col-md-4]=\"field.width == 1\" [class.col-md-10]=\"field.width == 2\">\r\n\t<div [ngClass]=\"{'radio': field.horizontal}\">\r\n        <span *ngFor='let radioElement of field.radios' [ngClass]=\"{'radio': !field.horizontal}\">\r\n            <label>\r\n                <input #radio name=\"{{field.id}}\" [(ngModel)]=\"field.value\" (click)=\"field.updateValueFromGui(radioElement.key)\" [value]=\"radioElement.key.toString()\" [disabled]=\"field.isReadonly()\" type=\"radio\">\r\n                <span> {{radioElement.name}}&nbsp;&nbsp;</span>\r\n            </label>\r\n        </span>\r\n\t</div>\r\n</div>\r\n"
        })
    ], RadioGroupComponent);
    return RadioGroupComponent;
}());
export { RadioGroupComponent };
var RadioGroupField = /** @class */ (function (_super) {
    tslib_1.__extends(RadioGroupField, _super);
    function RadioGroupField() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._radios = [];
        return _this;
    }
    Object.defineProperty(RadioGroupField.prototype, "radios", {
        get: function () {
            return this._radios;
        },
        set: function (value) {
            this._radios = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RadioGroupField.prototype, "horizontal", {
        get: function () {
            return this._horizontal;
        },
        set: function (value) {
            this._horizontal = value;
        },
        enumerable: true,
        configurable: true
    });
    RadioGroupField.prototype.setValue = function (value) {
        if (value != null) {
            value = value.toString();
        }
        _super.prototype.setValue.call(this, value);
    };
    RadioGroupField.prototype.updateValueFromGui = function (value) {
        this.setValue(value.toString());
    };
    RadioGroupField.prototype.isReadonly = function () {
        return _super.prototype.isReadonly.call(this) && this.isInitialValueSet;
    };
    RadioGroupField.prototype.fromJson = function (modelFormField) {
        var e_1, _a;
        _super.prototype.fromJson.call(this, modelFormField);
        this._horizontal = true;
        if (modelFormField["alignment"] != null) {
            if (modelFormField["alignment"] === "vertical") {
                this._horizontal = false;
            }
        }
        if (modelFormField["radios"] != null) {
            try {
                for (var _b = tslib_1.__values(modelFormField["radios"]), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var radioModel = _c.value;
                    var radio = new Radio(this.getForm(), this.translationService);
                    radio.name = radioModel.name;
                    radio.key = radioModel.key;
                    if (radio.key === modelFormField["defaultKey"]) {
                        this.setValue(radio.key);
                    }
                    this.radios.push(radio);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
    };
    return RadioGroupField;
}(ValueField));
export { RadioGroupField };
