// path-framework/dist/esm2015/path-framework/form/field/radio/radio-group.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output } from "@angular/core";
import { Radio } from "./radio";
import { ValueField } from "../value-field";
let RadioGroupComponent = class RadioGroupComponent {
};
tslib_1.__decorate([
    Input("field"),
    Output("field")
], RadioGroupComponent.prototype, "field", void 0);
RadioGroupComponent = tslib_1.__decorate([
    Component({
        selector: "path-radio-group",
        template: "<path-form-field-label [label]=\"field.name\" [required]=\"field.required\"></path-form-field-label>\r\n<div [class.col-md-4]=\"field.width == 1\" [class.col-md-10]=\"field.width == 2\">\r\n\t<div [ngClass]=\"{'radio': field.horizontal}\">\r\n        <span *ngFor='let radioElement of field.radios' [ngClass]=\"{'radio': !field.horizontal}\">\r\n            <label>\r\n                <input #radio name=\"{{field.id}}\" [(ngModel)]=\"field.value\" (click)=\"field.updateValueFromGui(radioElement.key)\" [value]=\"radioElement.key.toString()\" [disabled]=\"field.isReadonly()\" type=\"radio\">\r\n                <span> {{radioElement.name}}&nbsp;&nbsp;</span>\r\n            </label>\r\n        </span>\r\n\t</div>\r\n</div>\r\n"
    })
], RadioGroupComponent);
export { RadioGroupComponent };
export class RadioGroupField extends ValueField {
    constructor() {
        super(...arguments);
        this._radios = [];
    }
    get radios() {
        return this._radios;
    }
    set radios(value) {
        this._radios = value;
    }
    get horizontal() {
        return this._horizontal;
    }
    set horizontal(value) {
        this._horizontal = value;
    }
    setValue(value) {
        if (value != null) {
            value = value.toString();
        }
        super.setValue(value);
    }
    updateValueFromGui(value) {
        this.setValue(value.toString());
    }
    isReadonly() {
        return super.isReadonly() && this.isInitialValueSet;
    }
    fromJson(modelFormField) {
        super.fromJson(modelFormField);
        this._horizontal = true;
        if (modelFormField["alignment"] != null) {
            if (modelFormField["alignment"] === "vertical") {
                this._horizontal = false;
            }
        }
        if (modelFormField["radios"] != null) {
            for (const radioModel of modelFormField["radios"]) {
                const radio = new Radio(this.getForm(), this.translationService);
                radio.name = radioModel.name;
                radio.key = radioModel.key;
                if (radio.key === modelFormField["defaultKey"]) {
                    this.setValue(radio.key);
                }
                this.radios.push(radio);
            }
        }
    }
}
