// path-framework/dist/esm2015/path-framework/page/element/label/page-label.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output } from "@angular/core";
import { PageElement } from "../page-element";
let PageLabelComponent = class PageLabelComponent {
};
tslib_1.__decorate([
    Input("pageLabel"),
    Output("pageLabel")
], PageLabelComponent.prototype, "pageLabel", void 0);
PageLabelComponent = tslib_1.__decorate([
    Component({
        selector: "path-page-label",
        template: "<div class=\"col-md-12\" [innerHTML]=\"pageLabel.value\"></div>"
    })
], PageLabelComponent);
export { PageLabelComponent };
export class PageLabel extends PageElement {
    constructor(app, pathService, translationService) {
        super(app);
    }
    get value() {
        return this._value;
    }
    onClick() {
        this.app.navigateBack();
    }
    fromJson(modelElement) {
        super.fromJson(modelElement);
        this.type = "pageLabel";
        if (modelElement["value"] != null) {
            this._value = modelElement["value"];
        }
    }
}
