// path-framework/dist/esm2015/path-framework/page/element/button-group/button-group.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output } from "@angular/core";
import { PageElement } from "../page-element";
let ButtonGroupComponent = class ButtonGroupComponent {
};
tslib_1.__decorate([
    Input("buttonGroup"),
    Output("buttonGroup")
], ButtonGroupComponent.prototype, "buttonGroup", void 0);
ButtonGroupComponent = tslib_1.__decorate([
    Component({
        selector: "path-button-group",
        template: "<span>\r\n    <div *ngIf='buttonGroup.mobilePageBreak' class=\"tile-group tile-group-divider tile hide-on-desktop\"></div>\r\n    <div *ngIf='buttonGroup.visible' class=\"tile-group\">\r\n            <span *ngFor='let pageElement of buttonGroup.buttons'>\r\n            <!-- Button -->\r\n            <span *ngIf='pageElement.type == \"button\" || pageElement.type == \"newButton\"'>\r\n                <path-button [button]=\"pageElement\"></path-button>\r\n            </span>\r\n            <!-- Back Button -->\r\n            <span *ngIf='pageElement.type == \"backbutton\"'>\r\n                <path-back-button [button]=\"pageElement\"></path-back-button>\r\n            </span>\r\n            <!-- Delete Button -->\r\n            <span *ngIf='pageElement.type == \"deleteButton\"'>\r\n                <path-page-delete-button [button]=\"pageElement\"></path-page-delete-button>\r\n            </span>\r\n            <!-- Link Button (deprecated: DownloadButton -->\r\n            <span *ngIf='pageElement.type == \"linkButton\" || pageElement.type == \"downloadButton\"'>\r\n                <path-link-button [button]=\"pageElement\"></path-link-button>\r\n            </span>\r\n        </span>\r\n    </div>\r\n    <div *ngIf='buttonGroup.mobilePageBreak' class=\"tile-group tile-group-divider tile hide-on-desktop\"></div>\r\n</span>"
    })
], ButtonGroupComponent);
export { ButtonGroupComponent };
export class ButtonGroup extends PageElement {
    constructor(app) {
        super(app);
        this._buttons = [];
        this._mobilePageBreak = false;
    }
    get buttons() {
        return this._buttons;
    }
    get mobilePageBreak() {
        return this._mobilePageBreak;
    }
    addButton(button) {
        this._buttons.push(button);
        let length = 0;
        for (const currentButton of this.buttons) {
            length += (currentButton.width >= 2 ? 2 : 1);
        }
        this._mobilePageBreak = length > 1 && this.buttons.length >= 2;
    }
    updateButtonBorders() {
        if (this._buttons.length === 1) {
            // single button has no special border
            this._buttons[0].cssGroupBorder = "";
        }
        else {
            for (const button of this._buttons) {
                button.cssGroupBorder = "tile-grouped-middle";
            }
            if (this._buttons.length >= 1) {
                // set first and last button border
                this._buttons[0].cssGroupBorder = "tile-grouped-left";
                this._buttons[this._buttons.length - 1].cssGroupBorder = "tile-grouped-right";
            }
        }
    }
    fromJson(modelElement) {
        super.fromJson(modelElement);
        this.type = "buttonGroup";
    }
}
