// path-framework/dist/esm5/path-framework/page/element/label/page-label.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output } from "@angular/core";
import { PageElement } from "../page-element";
var PageLabelComponent = /** @class */ (function () {
    function PageLabelComponent() {
    }
    tslib_1.__decorate([
        Input("pageLabel"),
        Output("pageLabel")
    ], PageLabelComponent.prototype, "pageLabel", void 0);
    PageLabelComponent = tslib_1.__decorate([
        Component({
            selector: "path-page-label",
            template: "<div class=\"col-md-12\" [innerHTML]=\"pageLabel.value\"></div>"
        })
    ], PageLabelComponent);
    return PageLabelComponent;
}());
export { PageLabelComponent };
var PageLabel = /** @class */ (function (_super) {
    tslib_1.__extends(PageLabel, _super);
    function PageLabel(app, pathService, translationService) {
        return _super.call(this, app) || this;
    }
    Object.defineProperty(PageLabel.prototype, "value", {
        get: function () {
            return this._value;
        },
        enumerable: true,
        configurable: true
    });
    PageLabel.prototype.onClick = function () {
        this.app.navigateBack();
    };
    PageLabel.prototype.fromJson = function (modelElement) {
        _super.prototype.fromJson.call(this, modelElement);
        this.type = "pageLabel";
        if (modelElement["value"] != null) {
            this._value = modelElement["value"];
        }
    };
    return PageLabel;
}(PageElement));
export { PageLabel };
