// path-framework/dist/esm5/path-framework/page/element/custom/custom-container.component.js
import * as tslib_1 from "tslib";
import { Component, ComponentFactoryResolver, Input, OnDestroy, OnInit, ViewChild } from "@angular/core";
import { CustomDirective } from "./custom.directive";
var CustomContainerComponent = /** @class */ (function () {
    function CustomContainerComponent(componentFactoryResolver) {
        this.componentFactoryResolver = componentFactoryResolver;
    }
    CustomContainerComponent.prototype.ngOnInit = function () {
        this.loadComponent();
    };
    CustomContainerComponent.prototype.ngOnDestroy = function () {
    };
    CustomContainerComponent.prototype.loadComponent = function () {
        var componentFactory = this.componentFactoryResolver.resolveComponentFactory(this.pageElement.typeClass);
        var viewContainerRef = this.customDirective.viewContainerRef;
        viewContainerRef.clear();
        var componentRef = viewContainerRef.createComponent(componentFactory);
        componentRef.instance.pageElement = this.pageElement;
    };
    CustomContainerComponent.ctorParameters = function () { return [
        { type: ComponentFactoryResolver }
    ]; };
    tslib_1.__decorate([
        Input()
    ], CustomContainerComponent.prototype, "pageElement", void 0);
    tslib_1.__decorate([
        ViewChild(CustomDirective, { static: true })
    ], CustomContainerComponent.prototype, "customDirective", void 0);
    CustomContainerComponent = tslib_1.__decorate([
        Component({
            selector: "path-custom-container-component",
            template: "\n        <ng-template path-custom-directive></ng-template>"
        })
    ], CustomContainerComponent);
    return CustomContainerComponent;
}());
export { CustomContainerComponent };
var CustomPageElement = /** @class */ (function () {
    function CustomPageElement() {
    }
    Object.defineProperty(CustomPageElement.prototype, "pageElement", {
        get: function () {
            return this._pageElement;
        },
        set: function (value) {
            this._pageElement = value;
        },
        enumerable: true,
        configurable: true
    });
    return CustomPageElement;
}());
export { CustomPageElement };
