// path-framework/dist/esm5/path-framework/page/element/element-list/element-list.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output } from "@angular/core";
import { PageElement } from "../page-element";
var ElementListComponent = /** @class */ (function () {
    function ElementListComponent() {
    }
    tslib_1.__decorate([
        Input("elementList"),
        Output("elementList")
    ], ElementListComponent.prototype, "elementList", void 0);
    ElementListComponent = tslib_1.__decorate([
        Component({
            selector: "path-element-list",
            template: "<span></span>"
        })
    ], ElementListComponent);
    return ElementListComponent;
}());
export { ElementListComponent };
var ElementList = /** @class */ (function (_super) {
    tslib_1.__extends(ElementList, _super);
    function ElementList(app, pathService, translationService) {
        return _super.call(this, app) || this;
    }
    ElementList.prototype.fromJson = function (modelElement) {
        _super.prototype.fromJson.call(this, modelElement);
        this.type = "elementList";
    };
    return ElementList;
}(PageElement));
export { ElementList };
