// path-framework/dist/esm5/path-framework/page/element/button-group/button-group.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output } from "@angular/core";
import { PageElement } from "../page-element";
var ButtonGroupComponent = /** @class */ (function () {
    function ButtonGroupComponent() {
    }
    tslib_1.__decorate([
        Input("buttonGroup"),
        Output("buttonGroup")
    ], ButtonGroupComponent.prototype, "buttonGroup", void 0);
    ButtonGroupComponent = tslib_1.__decorate([
        Component({
            selector: "path-button-group",
            template: "<span>\r\n    <div *ngIf='buttonGroup.mobilePageBreak' class=\"tile-group tile-group-divider tile hide-on-desktop\"></div>\r\n    <div *ngIf='buttonGroup.visible' class=\"tile-group\">\r\n            <span *ngFor='let pageElement of buttonGroup.buttons'>\r\n            <!-- Button -->\r\n            <span *ngIf='pageElement.type == \"button\" || pageElement.type == \"newButton\"'>\r\n                <path-button [button]=\"pageElement\"></path-button>\r\n            </span>\r\n            <!-- Back Button -->\r\n            <span *ngIf='pageElement.type == \"backbutton\"'>\r\n                <path-back-button [button]=\"pageElement\"></path-back-button>\r\n            </span>\r\n            <!-- Delete Button -->\r\n            <span *ngIf='pageElement.type == \"deleteButton\"'>\r\n                <path-page-delete-button [button]=\"pageElement\"></path-page-delete-button>\r\n            </span>\r\n            <!-- Link Button (deprecated: DownloadButton -->\r\n            <span *ngIf='pageElement.type == \"linkButton\" || pageElement.type == \"downloadButton\"'>\r\n                <path-link-button [button]=\"pageElement\"></path-link-button>\r\n            </span>\r\n        </span>\r\n    </div>\r\n    <div *ngIf='buttonGroup.mobilePageBreak' class=\"tile-group tile-group-divider tile hide-on-desktop\"></div>\r\n</span>"
        })
    ], ButtonGroupComponent);
    return ButtonGroupComponent;
}());
export { ButtonGroupComponent };
var ButtonGroup = /** @class */ (function (_super) {
    tslib_1.__extends(ButtonGroup, _super);
    function ButtonGroup(app) {
        var _this = _super.call(this, app) || this;
        _this._buttons = [];
        _this._mobilePageBreak = false;
        return _this;
    }
    Object.defineProperty(ButtonGroup.prototype, "buttons", {
        get: function () {
            return this._buttons;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "mobilePageBreak", {
        get: function () {
            return this._mobilePageBreak;
        },
        enumerable: true,
        configurable: true
    });
    ButtonGroup.prototype.addButton = function (button) {
        var e_1, _a;
        this._buttons.push(button);
        var length = 0;
        try {
            for (var _b = tslib_1.__values(this.buttons), _c = _b.next(); !_c.done; _c = _b.next()) {
                var currentButton = _c.value;
                length += (currentButton.width >= 2 ? 2 : 1);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        this._mobilePageBreak = length > 1 && this.buttons.length >= 2;
    };
    ButtonGroup.prototype.updateButtonBorders = function () {
        var e_2, _a;
        if (this._buttons.length === 1) {
            // single button has no special border
            this._buttons[0].cssGroupBorder = "";
        }
        else {
            try {
                for (var _b = tslib_1.__values(this._buttons), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var button = _c.value;
                    button.cssGroupBorder = "tile-grouped-middle";
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_2) throw e_2.error; }
            }
            if (this._buttons.length >= 1) {
                // set first and last button border
                this._buttons[0].cssGroupBorder = "tile-grouped-left";
                this._buttons[this._buttons.length - 1].cssGroupBorder = "tile-grouped-right";
            }
        }
    };
    ButtonGroup.prototype.fromJson = function (modelElement) {
        _super.prototype.fromJson.call(this, modelElement);
        this.type = "buttonGroup";
    };
    return ButtonGroup;
}(PageElement));
export { ButtonGroup };
