// path-framework/dist/esm2015/path-framework/form/field/checkbox/checkbox-group.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output } from "@angular/core";
import { ValueField } from "../value-field";
let CheckboxGroupComponent = class CheckboxGroupComponent {
};
tslib_1.__decorate([
    Input("field"),
    Output("field")
], CheckboxGroupComponent.prototype, "field", void 0);
CheckboxGroupComponent = tslib_1.__decorate([
    Component({
        selector: "path-checkbox-group",
        template: "<path-form-field-label [label]=\"field.name\" [required]=\"field.required\"></path-form-field-label>\r\n<div class=\"checkbox\" [class.col-md-4]=\"field.width == 1\" [class.col-md-10]=\"field.width == 2\">\r\n    <span *ngFor=\"let entry of field.entries\">\r\n        <label class=\"checkbox-inline\">\r\n            <input type=\"checkbox\" [(ngModel)]=\"entry.selected\" (ngModelChange)=\"field.doClick()\">\r\n            <span class=\"checkbox-label\"> {{entry.name}}&nbsp;&nbsp;</span>\r\n        </label>\r\n    </span>\r\n</div>\r\n"
    })
], CheckboxGroupComponent);
export { CheckboxGroupComponent };
export class CheckboxGroupField extends ValueField {
    constructor() {
        super(...arguments);
        this._entries = [];
    }
    get entries() {
        return this._entries;
    }
    set entries(value) {
        this._entries = value;
    }
    doClick() {
        const values = [];
        for (const entry of this.entries) {
            if (entry.selected) {
                values.push(entry.key);
            }
        }
        this.setValue(values);
    }
    setValue(values) {
        super.setValue(values);
        for (const entry of this.entries) {
            entry.selected = false;
        }
        for (const value of values) {
            for (const entry of this.entries) {
                if (entry.key === value) {
                    entry.selected = true;
                }
            }
        }
    }
    fromJson(modelFormField) {
        super.fromJson(modelFormField);
        for (const entryModel of modelFormField["data"]) {
            const entry = new CheckboxGroupEntry();
            entry.name = entryModel.name;
            entry.key = entryModel.key;
            entry.selected = false;
            this.entries.push(entry);
        }
    }
}
export class CheckboxGroupEntry {
    get key() {
        return this._key;
    }
    set key(value) {
        this._key = value;
    }
    get name() {
        return this._name;
    }
    set name(value) {
        this._name = value;
    }
    get selected() {
        return this._selected;
    }
    set selected(value) {
        this._selected = value;
    }
}
