// path-framework/dist/esm5/path-framework/form/field/number/number-field.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output } from "@angular/core";
import { ValueField } from "../value-field";
var NumberFieldComponent = /** @class */ (function () {
    function NumberFieldComponent() {
    }
    tslib_1.__decorate([
        Input("field"),
        Output("field")
    ], NumberFieldComponent.prototype, "field", void 0);
    NumberFieldComponent = tslib_1.__decorate([
        Component({
            selector: "path-numberfield",
            template: "<path-form-field-label [label]=\"field.name\" [required]=\"field.required\"></path-form-field-label>\r\n<div [class.col-md-4]=\"field.width == 1\" [class.col-md-10]=\"field.width == 2\">\r\n    <div class=\"input-group\">\r\n        <input min=\"{{field.min}}\" max=\"{{field.max}}\" step=\"{{field.getStep()}}\" *ngIf=\"field.height == null || field.height <= 1\" #model=\"ngModel\" [readonly]=\"field.isReadonly()\" [class.required-field]=\"(!model.valid &&!model.pristine && field.required)\" [required]=\"field.required\" [(ngModel)]=field.value type=\"number\" class=\"form-control\" placeholder=\"\">\r\n    </div>\r\n</div>\r\n"
        })
    ], NumberFieldComponent);
    return NumberFieldComponent;
}());
export { NumberFieldComponent };
var NumberField = /** @class */ (function (_super) {
    tslib_1.__extends(NumberField, _super);
    function NumberField() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(NumberField.prototype, "min", {
        get: function () {
            return this._min;
        },
        set: function (value) {
            this._min = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NumberField.prototype, "max", {
        get: function () {
            return this._max;
        },
        set: function (value) {
            this._max = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NumberField.prototype, "digits", {
        get: function () {
            return this._digits;
        },
        set: function (value) {
            this._digits = value;
        },
        enumerable: true,
        configurable: true
    });
    NumberField.prototype.getStep = function () {
        if (this._digits == null) {
            return "";
        }
        else if (this._digits === 0) {
            return "0";
        }
        // e.g. 3 => 0.001
        var step = "0.";
        for (var k = 0; k < this._digits - 1; k++) {
            step = step + "0";
        }
        return step + "1";
    };
    NumberField.prototype.fromJson = function (modelFormField) {
        _super.prototype.fromJson.call(this, modelFormField);
        if (modelFormField["min"] != null) {
            this.min = modelFormField["min"];
        }
        if (modelFormField["max"] != null) {
            this.max = modelFormField["max"];
        }
        if (modelFormField["digits"] != null) {
            this.digits = modelFormField["digits"];
        }
    };
    return NumberField;
}(ValueField));
export { NumberField };
